### PR TITLE
Add support to MIGraphX backend and FP16 precision

### PR DESCRIPTION
**Pull request by terU3760**
Add support to MIGraphX backend and FP16 precision. Both the support to MIGraphX backend and FP16 precision has been coded and tested. All the files are just organized as those in IPU and STC backends.

### CODE REVIEW

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
+import os
+import json
+import logging
+import copy
+
+import torch
+# import lego
```
Comment (JianzheXiao) [inline byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py:7]: nit: this line can be removed

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                    0,
+                    "is_fallback":
+                    False,
+                    "input_tensor_map":
+                    config['model_info']['input_shape'],
+                    "output_tensor_map":
+                    config['model_info']['outputs'],
+                    "compiled_model": [
+                        {
+                            "compiled_bs": config["workload"]["batch_sizes"][-1],
```
Comment (JianzheXiao) [inline byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py:125]: In line 75, we compiled multiple batch sizes, so I believe we should have multiple result here as well.

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            if not os.path.exists(model_path):
+                new_input = {}
+                for key, value in config['model_info']['input_shape'].items():
+                    if(('videobert' in config['model_info']['model']) and (key=='text')):
+                        new_input[key]=value
+                    else:
+                        new_input[key]=[value[0]*batch_size]+value[1:]
+                model = migraphx.parse_onnx(model_onnx_path,map_input_dims=new_input,default_dim_value=batch_size)
+
+                if('FP16' in config['model_info']['model_precision']):
```
Comment (JianzheXiao) [inline byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py:92]: Here we don't recommend to change the original model config json file, we could add a interactive question in MIGRAPHX.json and call get_interact_profile to query the answer see CPU as an example, or we could just set it as default

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            1, config['model_info']['input_shape'], self.input_type)
+        input_tensors = []
+        #convert tf/pytorch model to onnx model, tf works, not verify pytorch now.
+        model_onnx_path=model_path
+        if self.framework == "Tensorflow":
+            model_onnx_path = os.path.join(model_path,self.model_name+ '.onnx')
+            if not os.path.exists(model_onnx_path):
+                saved_to_onnx.savedmodel_to_onnx(model_path, model_onnx_path)
+        elif self.framework == "Pytorch":
+            raise NotImplementedError("MIGraphX backend for models of PyTorch framework has not been implemented yet.")
```
Comment (JianzheXiao) [inline byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py:73]: I see the file imported torch_to_onnx but didn't use it here. Is there any reason?

**Review by JianzheXiao — CHANGES_REQUESTED** [review]

**Comment by JianzheXiao** [comment]
Another comment is that we should remove files in model_zoo and workloads, because 'fp32' here represent the original model's precision. If we would like to compile it as precision fp16, we could just update the precision inthe result dic in compile()

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py (hunk 3 above)
Comment (terU3760) [reply byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py:92]: Do you mean it should not create new config json file for FP16 precision models? But the only place to specify the precision is in calling launch.py and passing the parameter as "resnet50-tf-fp16" to the option "--task". It is hard coded to load that json file whose name is, e.g. "resnet50-tf-fp16.json" in the function load_workload of the file workload_store.py. What do you mean by "should not create new config json file for FP16 precision models"?

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py (hunk 4 above)
Comment (terU3760) [reply byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py:77]: Have fixed it.

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py (hunk 1 above)
Comment (terU3760) [reply byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py]: Have fixed it.

File: byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py (hunk 2 above)
Comment (terU3760) [reply byte_mlperf/backends/MIGRAPHX/compile_backend_migraphx.py]: Have fixed it. Now we set "compiled_bs" to the list of all batch sizes and "compiled_obj" to the list of saved models' file path.

File: byte_mlperf/backends/MIGRAPHX/runtime_backend_migraphx.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            if( self.framework == "Tensorflow" or self.framework == "Onnx"):
+                onnx_file_path = segment['compiled_model'][0]['compiled_obj']
+                import migraphx
+                onnx_file_path_for_batch_size = onnx_file_path.rsplit("/",2)
+                onnx_file_path=os.path.join(onnx_file_path_for_batch_size[0],str(self.batch_size),onnx_file_path_for_batch_size[-1])
+                model=migraphx.load(onnx_file_path,format='msgpack')
+            elif self.framework == "Pytorch":
+                raise NotImplementedError("MIGraphX backend for models of PyTorch framework has not been implemented yet.")
+            else:
+                # original_model = onnx.load(segment['compiled_model'][0]['compiled_obj'])
```
[the remark this answers is not in the document]
Comment (terU3760) [reply byte_mlperf/backends/MIGRAPHX/runtime_backend_migraphx.py]: Have fixed this.

**Comment by JianzheXiao** [comment]
Expired

**Pull request closed without merging**